### ISSUE TITLE
Reading multiple molecules from mmcif

### ISSUE THREAD

**Issue by multimeric** [openbabel/openbabel#2458]
- [x] I believe this to be a bug with Open Babel
- [ ] This is a feature request

## Environment Information
Open Babel version: 3.1.0
Operating system and version: Ubuntu 21.04

## Expected Behavior
Iterating over the molecules in an mmcif with multiple chains (e.g. ligands) should yield a different molecule for each chain (asym_id)

## Actual Behavior
Iterating over the molecules in an mmcif with multiple chains (e.g. ligands) will only yield one single molecule which contains the atoms from all chains

## Steps to Reproduce

1. Download this file
[P04406.mmcif.txt](https://github.com/openbabel/openbabel/files/8403838/P04406.mmcif.txt)

2. From the same directory the above file is located in, run:
```python
from openbabel import pybel
molecules = list(pybel.readfile("mmcif", "P04406.mmcif.txt"))
print(len(molecules))
```

However if you look at the actual file, you'll note that there are clearly multiple molecules. For instance, in this part of the `_atom_site` loop, we move from `ATOM` to `HETATM`, with different `asym_id` (`A` -> `B`) which indicate a change in molecule:
```
ATOM   2530 O  O   . GLU A 1 335 ? -24.426    9.206      -5.546     1.0  88.86 ? 335 GLU A O   1 P04406 UNP 335 E 
ATOM   2531 C  CG  . GLU A 1 335 ? -21.013    7.997      -2.666     1.0  88.86 ? 335 GLU A CG  1 P04406 UNP 335 E 
ATOM   2532 C  CD  . GLU A 1 335 ? -21.190    6.513      -2.996     1.0  88.86 ? 335 GLU A CD  1 P04406 UNP 335 E 
ATOM   2533 O  OE1 . GLU A 1 335 ? -20.200    5.869      -3.409     1.0  88.86 ? 335 GLU A OE1 1 P04406 UNP 335 E 
ATOM   2534 O  OE2 . GLU A 1 335 ? -22.265    5.907      -2.767     1.0  88.86 ? 335 GLU A OE2 1 P04406 UNP 335 E 
ATOM   2535 O  OXT . GLU A 1 335 ? -23.030    10.937     -5.405     1.0  88.86 ? 335 GLU A OXT 1 P04406 UNP 335 E 
HETATM 2536 O  O7N ? NAD B 2 336 ? 19.418873  -12.557346 25.040537  1.00 27.94 ? 336 NAD B O7N 1 ?      ?   ?   ? 
HETATM 2537 C  C7N ? NAD B 2 336 ? 18.608522  -13.471519 25.032057  1.00 31.53 ? 336 NAD B C7N 1 ?      ?   ?   ? 
HETATM 2538 N  N7N ? NAD B 2 336 ? 18.030472  -13.804909 23.869556  1.00 25.93 ? 336 NAD B N7N 1 ?      ?   ?   ? 
HETATM 2539 C  C3N ? NAD B 2 336 ? 18.274796  -14.219821 26.318806  1.00 24.78 ? 336 NAD B C3N 1 ?      ?   ?   ? 
HETATM 2540 C  C2N ? NAD B 2 336 ? 17.020575  -14.796659 26.528671  1.00 32.34 ? 336 NAD B C2N 1 ?      ?   ?   ? 
HETATM 2541 C  C4N ? NAD B 2 336 ? 19.216839  -14.345295 27.339706  1.00 28.76 ? 336 NAD B C4N 1 ?      ?   ?   ? 
HETATM 2542 C  C5N ? NAD B 2 336 ? 18.908211  -15.041785 28.508135  1.00 25.66 ? 336 NAD B C5N 1 ?      ?   ?   ? 
```

## Additional

This seems to relate to #912, which has been closed and sounds resolved, but seemingly this isn't actually resolved at all.

**Comment by multimeric** [comment]
Interestingly it seems that the PDB reader has the same issue. If you have multiple different chains in the same file, openbabel will still only output one molecule.

**Comment by nbehrnd** [comment]
To qualify for the a `.cif` (smaller molecules) or a `.mmcif` file (for larger protein/DNA/RNA models), the file should equally contain information about the symmetries present in the unit cell, as well as about the three vectors defining the parallelepiped (i.e., cell lengths a, b, c and the enclosed angles \alpha, \beta, \gamma). Depending on the symmetry met, some of these values may be equal (e.g., cubic Bravais class).  Your file seems to miss them.

Because the data set seems to be the result of a simulation, do you have the option for an output in a different file format?  Given the chemistry (size of the model .and. that there are so many amino acids around), `.pdb` likely would be available, too.

Speaking of small molecule data (e.g., `.cif` from the COD like [this one](http://www.crystallography.net/cod/2005645.html)), the archive attached below provides an example converting a single entry.  There, once a `.sdf`  is in hand (with all molecules  _in one model_) the `--separate` flag converts them into _multiple models_ one `.sdf` container.  Later, these are split into individual `.sdf` files with `-m`:

```bash
obabel transformation.sdf -O separate_models.sdf --separate --unique
obabel separate_models.sdf -O single.sdf -m
```
(Note: use `--unique` with what you intend in subsequent processing in mind.)

Lacking some crucial information, the same approach were not reasonable for `P04406.mmcif`.

[example_cod.zip](https://github.com/openbabel/openbabel/files/8408123/example_cod.zip)

**Comment by multimeric** [comment]
> Because the data set seems to be the result of a simulation, do you have the option for an output in a different file format? Given the chemistry (size of the model .and. that there are so many amino acids around), .pdb likely would be available, too.

No, this research output was only provided in the mmcif format. My understanding was that all `.pdb` file data was being migrated to mmcif due to its advantages.

It seems you have some concerns about the suitability of the `.mmcif` format for this use case, but it is absolutely already in use, and further the molecule identity can easily be obtained from the `asym_id`. I don't see any issue here asides from the missing feature for openbabel to parse this correctly.

**Comment by nbehrnd** [comment]
Applying the same pattern as for the small molecule example

```bash
$ obabel P04406.mmcif -O probe.sdf 
==============================
*** Open Babel Warning  in PerceiveBondOrders
  Failed to kekulize aromatic bonds in OBMol::PerceiveBondOrders (title is AF-P04406-F1)

1 molecule converted
```
followed by 
```bash
$obabel probe.sdf -O separate_models.sdf --separate
28 molecules converted
```
yields some «molecules» with little resemblance to what chemistry on the bench offers:

![inspection_DW](https://user-images.githubusercontent.com/19709084/161509368-136c33b1-d626-4f69-88d4-79a860c438d7.png)

There should be a better way to iterate over individual molecules in `.mmcif`, perhaps a question suitable for [mattermodelling.se](https://mattermodeling.stackexchange.com/)?

[split.zip](https://github.com/openbabel/openbabel/files/8408212/split.zip)

**Comment by nbehrnd** [comment]
@multimeric 
I gave the data file a second chance for the couple of [Jmol](http://jmol.sourceforge.net/) and OpenBabel.  For one, the conversion into individual molecules works much better (a Zn, a large protein, two smaller ligand-like molecules) after the `.mmcif` file read by Jmol was exported as a `.pdb`.  This was then followed  by the split into separate models (`--separate`) and split into individual files (`-m`) with OpenBabel.

With focus of attention to the two smaller molecules, however, the parameters of your protein folding algorithm appear to suffer a complication.  Intramolecular distances _of some_ atoms of oxygen and phosphor seem to be considerably less than the sum of the van der Waals radii of the other atoms they bind to.  As an example:

![problematic_01](https://user-images.githubusercontent.com/19709084/161627872-98727d98-21de-47ab-9a9f-ea651a218b95.png)

In analogy to other programs (for small molecule `.cif` data) it is plausible for me OpenBabel encountered these bumps, was at least puzzled, and in consequence, stopped processing the input `.mmcif` file.

The .zip archive attached below documents the processing _en detail_.

[jmol_split.zip](https://github.com/openbabel/openbabel/files/8413070/jmol_split.zip)

**Comment by multimeric** [comment]
Okay well it's good that JMOL can do this, but since it can, I was hoping OpenBabel could do the same. I'm trying to do this using pure python, without user interference.

**Comment by nbehrnd** [comment]
I read your lines in the tune of «nice that Jmol is able to
contribute to a solution, but it does so only via a GUI».

If this interpretation were correct, my reply to this is /no/.  With
e.g, flags `-ions`, especially the GUI-less JmolData.jar (equally
distributed in the Jmol archive) may run from the CLI in large
batches from the CLI.

All you need is to identify the commands necessary and write a script
.spt (there is an interactive Jmol script documentation; here, you
only need to gravitate around the keywords `load` for file input and
`write` for the subsequent export.  The script editor [Jmol.jar GUI,
File -> Script Editor] eases to compose them and probe their action,
too.). Later use the script from the CLI in the pattern of 

java -jar Jmol.jar -ions water_script.spt

where `water_script.spt` may contain for-loops, if-clauses, arrays,
etc.  Just look for e.g., «JmolData.jar apparently doesn't report
InChI back to CLI» in the Jmol user list archive back in February
2022 when I faced an automation problem Robert Hanson knew how to
address the Jmol way.  This command may be called from bash, Python,
AWK, etc.

Root entry: https://sourceforge.net/p/jmol/mailman/message/37610564/

**Comment by fredrikw** [comment]
I'm by no means an expert on biomolecule data formats, but as far as I can tell, openbabel is retaining the asym_id as chain information but it doesn't split the molecule into pieces. My guess is that you don't always want to have them split, i.e. in the case of a protein with several chains.
But you may split them in python, call `molecules[0].OBMol.Separate()` and you will get a tuple with one OBMol per disconnected structure.
